mempool/reactor_test.go

```diff
@@ -685,3 +685,184 @@ func TestMempoolVectors(t *testing.T) {
+
+	reactors[0].redundancyControl.triggerAdjustment()
+	time.Sleep(1 * time.Second)
+
+	reactors[0].redundancyControl.mtx.RLock()
+	dupTx := reactors[0].redundancyControl.duplicateTxs
+	firstTimeTx := reactors[0].redundancyControl.firstTimeTxs
+	reactors[0].redundancyControl.mtx.RUnlock()
+
+	// Now the counters should be reset
+	require.Equal(t, int64(0), dupTx)
+	require.Equal(t, int64(0), firstTimeTx)
+}
+
+// Testing the disabled route between two nodes
+// AS the description of DOG in the issue:
+// The core idea of the protocol is the following.
+// Consider a node A that receives from node B a transaction
+// that it already has. Let's assume B itself had received
+// the transaction from C. The fact that A received from B
+// a transaction it already has means that there must exist
+// a cycle in the network topology.
+// Therefore, a tells B to stop sending transactions B would be getting from C
+// (i.e. A tells B to disable route C → A → B).
+// We then reduce the redundancy level forcing A to tell B to re-enable the routes.
+func TestDOGDisabledRoute(t *testing.T) {
+	config := cfg.TestConfig()
+	config.Mempool.DOGProtocolEnabled = true
+
+	// Put the interval to a higher value to make sure the values don't get reset
+	config.Mempool.DOGAdjustInterval = 35 * time.Second
+	reactors, _ := makeAndConnectReactors(config, 3, nil)
+
+	secondNodeID := reactors[1].Switch.NodeInfo().ID()
+	secondNode := reactors[0].Switch.Peers().Get(secondNodeID)
+	secondNodeFromThird := reactors[2].Switch.Peers().Get(secondNodeID)
+
+	thirdNodeID := reactors[2].Switch.NodeInfo().ID()
+	thirdNodeFromFirst := reactors[0].Switch.Peers().Get(thirdNodeID)
+
+	firstNodeID := reactors[0].Switch.NodeInfo().ID()
+	firstNodeFromThird := reactors[2].Switch.Peers().Get(firstNodeID)
+
+	// create random transactions
+	txs := NewRandomTxs(numTxs, 20)
+	// Making sure they are unique
+	txUnique := make(map[string][]byte, len(txs))
+	for _, tx := range txs {
+		txUnique[string(tx.Hash())] = tx
+	}
+
+	// Add transactions to node 3 from node 2
+	// node3.senders[tx] = node2
+	for _, tx := range txUnique {
+		_, err := reactors[2].TryAddTx(tx, secondNodeFromThird)
+		require.NoError(t, err)
+	}
+
+	// Add the same transactions to node 1 from node 2
+	for _, tx := range txUnique {
+		_, err := reactors[0].TryAddTx(tx, secondNode)
+		require.NoError(t, err)
+	}
+
+	// Trying to add the same transactions node 1 has received
+	// from node 2, but this time from node 3
+	// Node 1 should now ask node 3 to disable the route between
+	// a node that has sent this tx to node 3(node 2) and node1
+	for _, tx := range txUnique {
+		_, err := reactors[0].TryAddTx(tx, thirdNodeFromFirst)
+		// The transaction is in cache, hence the Error
+		require.ErrorIs(t, err, ErrTxInCache)
+	}
+
+	reactors[0].redundancyControl.triggerAdjustment()
+	// Wait for the redundancy adjustment to kick in
+	time.Sleep(1 * time.Second)
```

Comment:
This can be replaced by `reactors[0].redundancyControl.adjustRedundancy(reactors[0])`.